src/dfx-core/src/extension/manifest/custom_canister_type.rs

```diff
@@ -0,0 +1,364 @@
+#[derive(Clone, Serialize, Deserialize, Debug)]
+struct Replace {
+    input: String,
+    search: String,
+    output: String,
+}
+
+#[derive(Debug)]
+enum OpError {
+    InvalidTemplate(handlebars::RenderError),
+    InvalidReplace(regex::Error),
+}
+
+impl std::fmt::Display for OpError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            OpError::InvalidTemplate(e) => write!(f, "Invalid template: {}", e),
+            OpError::InvalidReplace(e) => write!(f, "Invalid regex: {}", e),
+        }
+    }
+}
+
+impl CustomCanisterTypeDeclaration {
+    fn apply_template(
+        &self,
+        values: BTreeMap<FieldName, JsonValue>,
+    ) -> Result<JsonMap<FieldName, JsonValue>, OpError> {
+        let mut remove_fields = vec![];
+        let mut final_fields = JsonMap::new();
+        for (field_name, op) in self
+            .0
+            .clone()
+            .into_iter()
+            .collect::<Vec<_>>()
+            .clone()
+            .into_iter()
+        {
+            match op {
+                Op::NumberValue(x) => {
+                    final_fields.insert(field_name, x.into());
+                }
+                Op::BoolValue(x) => {
+                    final_fields.insert(field_name, x.into());
+                }
+
+                Op::Template(template) => {
+                    let x = Handlebars::new()
+                        .render_template(&template, &values)
+                        .map_err(OpError::InvalidTemplate)?;
+                    final_fields.insert(field_name, x.into());
+                }
+                Op::Replace { replace } => {
+                    let input = Handlebars::new()
+                        .render_template(&replace.input, &values)
+                        .map_err(OpError::InvalidTemplate)?;
+                    let re = Regex::new(&replace.search).map_err(OpError::InvalidReplace)?;
+                    let x = re.replace_all(&input, &replace.output).to_string();
+                    final_fields.insert(field_name, x.into());
+                }
+                Op::Remove { remove } if remove => {
+                    remove_fields.push(field_name);
+                }
+                _ => {}
+            }
+        }
+        // Removing fields should be done last because of the order of the fields in the map.
+        // It's easier to do in second for loop than to sort Ops beforehand, bacause Op would need to implement PartialOrd,
+        // which is not possible, because serde_json::Number does not implement it.
+        for field_name in remove_fields {
+            final_fields.remove(&field_name);
+        }
+        // Override custom canister declaration values by the real canister_declaration
+        for (key, value) in values.iter() {
+            if key != "type" && key != "canister_name" {
+                final_fields.insert(key.clone(), value.clone());
+            }
+        }
```

Comment:
@ericswanson-dfinity I just realized this is buggy... I'm working on a fix

Reply:
After all, I'm not sure what should be the desired behavior...
I extended unit tests with:
```rust
    #[test]
    fn test_op_replace_basic() {
        test_op!(
            custom_canister_template = r#"
        {
            "main": { "replace": { "input": "{{canister_name}}", "search": "frontend_(.*)", "output": "oowee/$1/main.ts" } }
        }
        "#,
            dfx_json_canister_values = r#"
        {
            "canister_name": "frontend_xyz"
        }
        "#,
            expected = r#"
        {
            "main": "oowee/xyz/main.ts"
        }
        "#
        );
    }

    #[test]
    fn test_op_replace_nested() {
        test_op!(
            custom_canister_template = r#"
        {
            "main": { "replace": { "input": "{{main}}", "search": "(.*)", "output": "oowee/$1" } }
        }
        "#,
            dfx_json_canister_values = r#"
        {
            "main": "src/main.ts"
        }
        "#,
            expected = r#"
        {
            "main": "oowee/src/main.ts"
        }
        "#
        );
    }
```

the `_basic` will pass, but what should be the expected outcome for `_nested`? This is related to https://github.com/dfinity/sdk/pull/3222#issuecomment-1624073606

Reply:
I don't understand how this is related to the comment.

From reading the comment I think the expectation is that
``` json
{
    "type": "kybra",
    "gzip": "true"
}
```
should transform according to the rules of the kybra extension, but also keep the "gzip" field.

Also I think these tests would benefit from some prose descriptions of what's being tested, what's the expected behavior, and what are the nuances.  As a reader I don't want to have to parse a regex in my head in order to figure out what the test is checking.

Reply:
apologies, I mixed too much info in that example; let me untangle
### case 1 - append user-defined field
is when someone defines `gzip: true` in their dfx.json, and `gzip` is not defined in custom canister type template. This works as expected. 
```rust
    #[test]
    fn test_op_replacement_1() {
        test_op!(
            custom_canister_template = r#"
        {
            "main": "something.py"
        }
        "#,
            dfx_json_canister_values = r#"
        {
            "gzip": true
        }
        "#,
            expected = r#"
        {
            "gzip": true,
            "main": "something.py"
        }
        "#
        );
    }
```

### case 2 - override the field
is when someone defines `gzip: true` in their dfx.json, and `gzip: false` is defined in custom canister type template. This also works as expected. 

```rust
    #[test]
    fn test_op_replacement_2() {
        test_op!(
            custom_canister_template = r#"
        {
            "main": "something.py",
            "gzip": false
        }
        "#,
            dfx_json_canister_values = r#"
        {
            "gzip": true
        }
        "#,
            expected = r#"
        {
            "gzip": true,
            "main": "something.py"
        }
        "#
        );
    }
```

### case 3 - nested field override 
is when someone defines `gzip: true` in their dfx.json, and `gzip: false` is defined in custom canister type template, and also `{{gzip}}` is used in the template
```rust 
    #[test]
    fn test_op_replacement_3() {
        test_op!(
            custom_canister_template = r#"
        {
            "main": "{{gzip}}.py",
            "gzip": false
        }
        "#,
            dfx_json_canister_values = r#"
        {
            "gzip": true
        }
        "#,
            expected = r#"
        {
            "gzip": true,
            "main": "true.py"
        }
        "#
        );
    }
```
Currently, the above test will pass, which I think makes the most sense?